File: 2024/Day9/main.cpp
#include <algorithm>
#include <charconv>
#include <fstream>
#include <ranges>
#include <string>
#include <string_view>
#include <vector>
#include <cstdio>
#include <unordered_map>
#include <numeric>
#include <print>

namespace r = std::ranges;
namespace v = std::ranges::views;
using namespace std::string_view_literals;
using namespace std::string_literals;

constexpr bool SAMPLE_DATA = false;

std::string read_file(std::string_view filename) {
    std::ifstream file(filename.data());
    file.seekg(0, std::ios::end);
    size_t size = file.tellg();
    std::string contents{};
    contents.reserve(size);
    file.seekg(0, std::ios::beg);
    contents.assign((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    return contents;
}

int64_t part1(std::vector<int64_t> filesystem) {
    size_t i = filesystem.size() - 1;
    size_t latest_idx = 0;
    for (const auto& c : filesystem | v::reverse) {
        if (auto it = r::find(filesystem.begin() + latest_idx, filesystem.end(), -1); it != filesystem.end()) {
            latest_idx = std::distance(filesystem.begin(), it);
            filesystem[latest_idx] = c;
            filesystem[i] = -2;
            i -= 1;
        } else {
            break;
        }
    }
    int64_t tot = 0;
    for (auto&& [i, c] : filesystem | v::enumerate) {
        if (c >= 0) {
            tot += (i * c);
        } else {
            break;
        }
    }
    return tot;
} 

int64_t part2(std::vector<int64_t> filesystem) {
    int64_t curtype = filesystem[0];
    auto spaces = curtype == -1 ? std::vector<std::pair<int64_t, int64_t>>{{0, 0}} : std::vector<std::pair<int64_t, int64_t>>{};
    auto files = curtype != -1 ? std::vector<std::pair<int64_t, int64_t>>{{0, 0}} : std::vector<std::pair<int64_t, int64_t>>{};
    for (auto&& [i, c] : filesystem | v::enumerate) {
        if (curtype == -1) {
            if (c == -1) {
                spaces.back().second += 1;
            } else {
                curtype = c;
                files.push_back({i, 1});
            }
        } else {
            if (c == -1) {
                curtype = -1;
                spaces.push_back({i, 1});
            } else if (c == curtype) {
                files.back().second += 1;
            } else {
                curtype = c;
                files.push_back({i, 1});
            }
        }
    }
    for (auto&& [n, f] : v::zip(v::iota(size_t{0}, files.size()) | v::reverse, files | v::reverse)) {
        auto&& [file_idx, file_size] = f;
        for (auto&& [i, s] : spaces | v::enumerate) {
            auto&& [space_idx, space_size] = s;
            if (space_idx > file_idx) break;
            if (file_size <= space_size) {
                r::fill_n(filesystem.begin() + space_idx, file_size, n);
                r::fill_n(filesystem.begin() + file_idx, file_size, -1);
                spaces[i] = {space_idx + file_size, space_size - file_size};
                break;
            }
        }
    }
    int64_t tot = 0;
    for (auto&& [i, c] : filesystem | v::enumerate) {
        if (c != -1) {
            tot += (i * c);
        }
    }
    return tot;
}

int main() {
    auto data = read_file(SAMPLE_DATA ? "sample_data.txt"sv : "data.txt"sv);
    std::vector<int64_t> fs{};
    fs.reserve(data.size());
    for (auto&& [i, f] : data | v::chunk(2) | v::enumerate) {
        auto fit = f.begin();
        r::fill_n(std::back_inserter(fs), (size_t)(*fit - '0'), i);
        if (++fit != f.end()) {
            r::fill_n(std::back_inserter(fs), (size_t)(*fit - '0'), -1);
        }
    }
    std::println("{} {}", part1(fs), part2(fs));
}

File: 2024/Day9/main.py
import itertools
import copy

SAMPLE_DATA: bool = False
if SAMPLE_DATA:
    filename = "sample_data.txt"
else:
    filename = "data.txt"


with open(filename, 'r') as f:
    data = [int(c) for c in f.read()]

def part1(filesystem: list[str]):
    i = len(filesystem) - 1
    latest_idx = 0
    for c in reversed(filesystem):
        try:
            latest_idx = filesystem.index('.', latest_idx)
            filesystem[latest_idx] = c    
            filesystem[i] = '#'
            i -= 1
        except ValueError:
            break
    tot = 0
    for i, c in enumerate(filesystem):
        if c.isdigit():
            tot += (i * int(c))
        else:
            break
    return tot

def part2(filesystem: list[str]):
    curtype = f if (f := filesystem[0]) != '.' else -1
    spaces = [(0, 0)] if curtype == -1 else []
    files = [(0, 0)] if curtype != -1 else []
    for i, c in enumerate(filesystem):
        if curtype == -1:
            if c == '.':
                idx, size = spaces[-1]
                spaces[-1] = (idx, size + 1)
            else:
                curtype = c
                files.append((i, 1))
        else:
            if c == '.':
                curtype = -1
                spaces.append((i, 1))
            elif c == curtype:
                idx, size = files[-1]
                files[-1] = (idx, size + 1)
            else:
                curtype = c
                files.append((i, 1))
    for n, (file_idx, file_size) in zip(reversed(range(len(files))), reversed(files)):
        for i, (space_idx, space_size) in enumerate(spaces):
            if space_idx > file_idx: # file is before space
                break
            if file_size <= space_size: # file can fit in space
                for j in range(space_idx, space_idx + file_size):
                    filesystem[j] = str(n)
                for j in range(file_idx, file_idx + file_size):
                    filesystem[j] = '.'
                spaces[i] = (space_idx + file_size, space_size - file_size)
                break
    tot = 0
    for i, c in enumerate(filesystem):
        if c.isdigit():
            tot += (i * int(c))
    return tot

        

arranged: list[str] = []

for i, f in enumerate(itertools.batched(data, 2)):
    arranged.extend(itertools.repeat(str(i), f[0]))
    if len(f) == 2:
        arranged.extend(itertools.repeat('.', f[1]))
print(f'{part1(copy.deepcopy(arranged))} {part2(arranged)}')